doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,697 @@
+
+   This document is a draft and a work in progress. Updates to it based
+   on experience with optimization are welcome!
+
+
+A Few Words About Process
+-------------------------
+
+Before we dive into optimization techniques, let's briefly discuss some
+good programming practices when working with Chapel. Optimization efforts
+will involve changing the code, and optimizations that cause the program
+to stop working correctly aren't any use at all. It needs to be easy to
+check that everything is working.
+
+Revision Control
+~~~~~~~~~~~~~~~~
+
+It's important to use revision control tools (e.g. ``git``). Revision
+control helps with optimization specifically:
+
+ * You can go back to an earlier version if your changes break the
+   program or make it slower
+
+ * You can record your progress towards improving performance
+
+Testing
+~~~~~~~
+
+It's important to have tests that are easy to run in order to check that
+your program is working correctly.
+
+We have used the following steps when testing a new change to a parallel
+and distributed Chapel application:
+
+ * Compile with checking enabled (i.e. without ``--fast``)
+
+   * Test with 1 locale and ``--dataParTasksPerLocale=1``
+
+   * Gradually increase ``--dataParTasksPerLocale`` until you are
+     satisfied it is working. At a minimum, try leaving this argument off
+     to run with number of tasks == number of cores.
+
+   * Test while simulating multiple locales on your development system
+     (see the next section)
+
+     Gradually increase the number of locales and the number of tasks
+     until you are satisfied it is working.
+
+You can compile with ``--fast`` and measure performance before or after
+this process. If you are targeting distributed memory, measure
+performance locally oversubscribed (see below) or by running on a
+multi-node system.
+
+Simulating Multiple Locales on a Laptop
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It's possible to run multiple processes on your laptop or workstation
+that simulate the communication that happens on a big multi-node
+system. This is also called running **locally oversubscribed**.
+
+In this setting, the program will generally run much slower than it would
+when running in a multi-node system. As a result, you might need to run
+with a smaller problem size to test in this way. Nonetheless, it is
+possible to identify and optimize communication.  In communication-bound
+applications, optimizing the communication can make the locally
+oversubscribed program run faster.  Another important technique is, when
+working locally oversubscribed, try to reduce the communication counts
+(communication counts are discussed below).
+The communication counts will match (or come close to matching) between
+running locally oversubscribed and running on a multi-node system with
+the same number of locales.  Reductions in communication counts will
+generally lead to improvements in the scalability of your application on
+multiple nodes.
```

Comment:
What would you think about adding a note here about this reduction having to be a significant one to improve performance? I am a little worried of users getting obsessed with reducing comm counts by a handful, which would probably doesn't impact actual multinode performance as much as they would like.

Reply:
Changed it to

> Significant reductions in communication counts will generally lead to improvements in the scalability of your application on multiple nodes.